src/connected/autobinding.ts

```diff
@@ -0,0 +1,137 @@
+/* --------------------------------------------------------------------------------------------
+ * SonarLint for VisualStudio Code
+ * Copyright (C) 2017-2022 SonarSource SA
+ * sonarlint@sonarsource.com
+ * Licensed under the LGPLv3 License. See LICENSE.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+'use strict';
+
+import * as VSCode from 'vscode';
+import { BindingService } from './binding';
+import { ConnectionSettingsService } from '../settings/connectionsettings';
+
+const AUTOBINDING_THRESHOLD = 5;
+const ATTEMPT_AUTOBINDING_ACTION = 'Attempt Auto-binding';
+const CHOOSE_MANUALLY_ACTION = 'Choose Manually';
+const DONT_ASK_AGAIN_ACTION = "Don't ask again";
+export const DO_NOT_ASK_ABOUT_AUTO_BINDING_FLAG = 'doNotAskAboutAutoBinding';
+const DEFAULT_CONNECTION_ID = '<default>';
+
+export class AutoBindingService {
+  private static _instance: AutoBindingService;
+
+  static init(
+    bindingService: BindingService,
+    workspaceState: VSCode.Memento,
+    settingsService: ConnectionSettingsService
+  ): void {
+    AutoBindingService._instance = new AutoBindingService(bindingService, workspaceState, settingsService);
+  }
+
+  constructor(
+    private readonly bindingService: BindingService,
+    private readonly workspaceState: VSCode.Memento,
+    private readonly settingsService: ConnectionSettingsService
+  ) {}
+
+  static get instance(): AutoBindingService {
+    return AutoBindingService._instance;
+  }
+
+  async autoBindAllUnboundFolders() {
+    if (this.workspaceState.get(DO_NOT_ASK_ABOUT_AUTO_BINDING_FLAG)) {
+      return;
+    }
+    const unboundFolders = VSCode.workspace.workspaceFolders.filter(workspaceFolder =>
+      this.bindingService.shouldBeAutoBound(workspaceFolder)
+    );
+    if (unboundFolders.length > AUTOBINDING_THRESHOLD) {
+      this.askUserBeforeAutoBinding();
+    }
+    unboundFolders.forEach(unboundFolder => {
+      this.autoBindFolder(unboundFolder);
+    });
```

Comment:
I believe this should be in an `else` block: the way I understand the code, this path will be executed after the "too many folders path", regardless of whether there are indeed too many folders or not (confirmed by manual test).